orttraining/orttraining/test/training_api/test_runner.cc

```diff
@@ -0,0 +1,323 @@
+  std::string tag("train");
+
+  for (size_t epoch = 0, batch_idx = 0; epoch < NUM_EPOCHS; ++epoch) {
+    for (auto it = data_loader.begin(); it != data_loader.end(); ++it) {
+      std::vector<OrtValue>& inputs = *it;
+      std::vector<OrtValue> fetches;
+      ORT_ENFORCE(module.TrainStep(inputs, fetches).IsOK());
+
+      const Tensor& loss_tensor = fetches[3].Get<Tensor>();
+
+      tensorboard->AddSummary(*(loss_tensor.template Data<std::string>()), batch_idx, tag);
+      std::cout << "Batch # : " << batch_idx << " Loss: " << GetLossValue(loss_tensor) << std::endl;
+
+      if (batch_idx % GRAD_ACC_STEPS == 0) {
+        // gradient accumulation steps completed
+        ORT_ENFORCE(optimizer.Step().IsOK());
+        // modify learning rate
+        ORT_ENFORCE(scheduler->Step().IsOK());
+        ORT_ENFORCE(optimizer.ResetGrad().IsOK());
+      }
+
+      if (batch_idx % EVAL_STEPS == 0) {
+        std::vector<OrtValue> eval_results;
+        ORT_ENFORCE(module.EvalStep(inputs, eval_results).IsOK());
+      }
+
+      if (batch_idx % SAVE_STEPS == 0) {
+        // save trained weights
+        utils::CheckpointStates state_dicts_to_save;
+        ORT_ENFORCE(module.GetStateDict(state_dicts_to_save.named_parameters).IsOK());
+        ORT_ENFORCE(optimizer.GetStateDict(state_dicts_to_save.optimizer_states).IsOK());
+        std::string ckpt_file = params.output_dir + "/resnet50_" + std::to_string(batch_idx) + ".ckpt";
+        ORT_ENFORCE(utils::Ort_Save(state_dicts_to_save, ckpt_file).IsOK());
+      }
+
+      batch_idx++;
+    }
+  }
+
+  return Status::OK();
+}
+
+#define RETURN_IF_FAIL(expr)                                \
+  do {                                                      \
+    auto status = (expr);                                   \
+    if ((!status.IsOK())) {                                 \
+      printf("Fail: %s \n", status.ErrorMessage().c_str()); \
+      return -1;                                            \
+    }                                                       \
+  } while (0);
+
+int main(int argc, char* argv[]) {
+  TestRunnerParameters params;
+  OrtTestRunnerParameters ort_params{};
+  RETURN_IF_FAIL(ParseArguments(argc, argv, params, ort_params));
+
+  // setup logger, be noted: LOGS_DEFAULT must be after logging manager initialization.
+  string default_logger_id{"Default"};
+  logging::LoggingManager default_logging_manager{std::make_unique<logging::CLogSink>(),
+                                                  ort_params.log_severity,
+                                                  false,
+                                                  logging::LoggingManager::InstanceType::Default,
+                                                  &default_logger_id,
+                                                  ort_params.vlog_level};
+#ifdef USE_CUDA
+  OrtCUDAProviderOptions provider_options{};
+  provider_options.do_copy_in_default_stream = true;
```

Comment:
curious why we are not going with the default here

Reply:
should be fine, because do_copy_in_default_stream is True by default now.